src/rt/monitor_.d

```diff
@@ -1,444 +1,289 @@
+// This is what the monitor reference in Object points to
+alias IMonitor = Object.Monitor;
+alias DEvent = void delegate(Object);
 
-version( Windows )
+version (Windows)
 {
-    static __gshared CRITICAL_SECTION _monitor_critsec;
-
-    extern (C) void _STI_monitor_staticctor()
+    version (CRuntime_DigitalMars)
     {
-        debug(PRINTF) printf("+_STI_monitor_staticctor()\n");
-        if (!inited)
-        {
-            InitializeCriticalSection(&_monitor_critsec);
-            inited = 1;
-        }
-        debug(PRINTF) printf("-_STI_monitor_staticctor()\n");
+        pragma(lib, "snn.lib");
     }
-
-    extern (C) void _STD_monitor_staticdtor()
+    else version (CRuntime_Microsoft)
     {
-        debug(PRINTF) printf("+_STI_monitor_staticdtor() - d\n");
-        if (inited)
-        {
-            inited = 0;
-            DeleteCriticalSection(&_monitor_critsec);
-        }
-        debug(PRINTF) printf("-_STI_monitor_staticdtor() - d\n");
+        pragma(lib, "libcmt.lib");
+        pragma(lib, "oldnames.lib");
     }
+    import core.sys.windows.windows;
 
-    extern (C) void _d_monitor_create(Object h)
+    struct Monitor
     {
-        /*
-         * NOTE: Assume this is only called when h.__monitor is null prior to the
-         * call.  However, please note that another thread may call this function
-         * at the same time, so we can not assert this here.  Instead, try and
-         * create a lock, and if one already exists then forget about it.
-         */
-
-        debug(PRINTF) printf("+_d_monitor_create(%p)\n", h);
-        assert(h);
-        Monitor *cs;
-        EnterCriticalSection(&_monitor_critsec);
-        if (!h.__monitor)
-        {
-            cs = cast(Monitor *)calloc(Monitor.sizeof, 1);
-            assert(cs);
-            InitializeCriticalSection(&cs.mon);
-            setMonitor(h, cs);
-            cs.refs = 1;
-            cs = null;
-        }
-        LeaveCriticalSection(&_monitor_critsec);
-        if (cs)
-            free(cs);
-        debug(PRINTF) printf("-_d_monitor_create(%p)\n", h);
+        IMonitor impl; // for user-level monitors
+        DEvent[] devt; // for internal monitors
+        size_t refs; // reference count
+        CRITICAL_SECTION mtx;
     }
 
-    extern (C) void _d_monitor_destroy(Object h)
-    {
-        debug(PRINTF) printf("+_d_monitor_destroy(%p)\n", h);
-        assert(h && h.__monitor && !getMonitor(h).impl);
-        DeleteCriticalSection(&getMonitor(h).mon);
-        free(h.__monitor);
-        setMonitor(h, null);
-        debug(PRINTF) printf("-_d_monitor_destroy(%p)\n", h);
-    }
+    alias Mutex = CRITICAL_SECTION;
+
+    alias initMutex = InitializeCriticalSection;
+    alias destroyMutex = DeleteCriticalSection;
+    alias lockMutex = EnterCriticalSection;
+    alias unlockMutex = LeaveCriticalSection;
+}
+else version (Posix)
+{
+    import core.sys.posix.pthread;
 
-    extern (C) void _d_monitor_lock(Object h)
+    struct Monitor
```

Comment:
With the help of alias Mutex, this struct could be moved out of the conditional.